### PR TITLE
Grok

### PR DESCRIPTION
**Pull request by intellinjun**
## Type of Change

feature or bug fix or documentation or others
API changed or not

## Description
enable grok
detail description 
[Grok-314b](https://huggingface.co/keyfan/grok-1-hf)
- Tanh in mha

### CODE REVIEW

**Comment by intellinjun** [comment]
image
grok-q40-gguf result

File: neural_speed/core/ne_layers.c [hunk 1]
```diff
@@ -707,6 +709,9 @@ static inline int ne_up(int n, int m) {
   return (n + m - 1) & ~(m - 1);
 }
 
+// static inline  void ne_vec_tanh_f32 (const int n, float * y, const float * x) { for (int i = 0; i < n; ++i) y[i] =
```
Comment (zhentaoyu) [inline neural_speed/core/ne_layers.c:712]: why disable this function?

File: neural_speed/models/grok/grok_utils.cpp [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+  vocab_only = vocab_only_;
+  auto& model = lctx.model;
+  ml.reset(new model_model_loader(path_model, use_mmap, vocab_only));
+  lctx.vocab = std::move(ml->file_loaders.at(0)->vocab);
+  model.hparams = ml->file_loaders.at(0)->hparams;
+  model_file_version file_version = ml->file_loaders.at(0)->file_version;
+  auto& hparams = model.hparams;
+  n_ff = hparams.ffn_hidden_size;
+  fprintf(stderr, "%s: n_vocab    = %u\n", __func__, hparams.n_vocab);
+  fprintf(stderr, "%s: n_ctx      = %u\n", __func__, hparams.max_seq_len);
```
Comment (zhentaoyu) [inline neural_speed/models/grok/grok_utils.cpp:68]: `lctx.n_ctx`?

**Review by zhentaoyu — APPROVED** [review]

**Review by a32543254 — APPROVED** [review]
LGTM

File: neural_speed/core/ne_layers.c (hunk 1 above)
Comment (intellinjun) [reply neural_speed/core/ne_layers.c:712]: This function is redefined, I will remove it.

File: neural_speed/models/grok/grok.h [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+enum grok_model {
+  GROK_314B,
+};
+
+static const model_scratch grok_mem_req(int n_layers, float enlarge_scale = 1.0f) {
+  switch (n_layers) {
+    case 64:
+      return {
+          static_cast<unsigned long long>(enlarge_scale * 3072) * MB,
```
Comment (Zhenzhong1) [inline neural_speed/models/grok/grok.h:29]: static_cast<unsigned long long>(scratch_size_ratio * 4096) * MB,
          static_cast<unsigned long long>(scratch_size_ratio * 2048) * MB,
          static_cast<unsigned long long>(scratch_size_ratio * 4096 * 10) * MB,

can we use this policy to keep the same as other models?

File: neural_speed/models/grok/grok_utils.cpp [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+  fprintf(stderr, "%s: n_embd     = %u\n", __func__, hparams.n_embd);
+  fprintf(stderr, "%s: n_mult     = %u\n", __func__, hparams.n_mult);
+  fprintf(stderr, "%s: n_head     = %u\n", __func__, hparams.n_head);
+  fprintf(stderr, "%s: n_head_kv  = %u\n", __func__, hparams.n_head_kv);
+  fprintf(stderr, "%s: n_layer    = %u\n", __func__, hparams.n_layer);
+  fprintf(stderr, "%s: n_rot      = %u\n", __func__, hparams.n_rot);
+  fprintf(stderr, "%s: n_embd_head_k     = %u\n", __func__, hparams.n_embd_head_k);
+  fprintf(stderr, "%s: n_ff       = %u\n", __func__, n_ff);
+  fprintf(stderr, "%s: n_parts    = %zu\n", __func__, ml->file_loaders.size());
```
Comment (Zhenzhong1) [inline neural_speed/models/grok/grok_utils.cpp:77]: please remove unused parameters

File: neural_speed/models/grok/grok_utils.cpp [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+  n_embd = hparams.n_embd;
+  n_vocab = hparams.n_vocab;
+  n_layer = hparams.n_layer;
+  n_head_kv = hparams.n_head_kv;
+  n_embd_head_k = hparams.n_embd_head_k;
+  n_head = hparams.n_head;
+  n_expert = hparams.n_experts;
+  n_expert_used = hparams.n_experts_used;
```
Comment (Zhenzhong1) [inline neural_speed/models/grok/grok_utils.cpp:85]: please print these used parameters.

File: neural_speed/models/grok/grok_utils.cpp (hunk 4 above)
Comment (intellinjun) [reply neural_speed/models/grok/grok_utils.cpp:75]: down

File: neural_speed/models/grok/grok_utils.cpp (hunk 5 above)
Comment (intellinjun) [reply neural_speed/models/grok/grok_utils.cpp:82]: done

File: neural_speed/models/grok/grok.h (hunk 3 above)
Comment (intellinjun) [reply neural_speed/models/grok/grok.h:29]: What's mean?

Comment (Zhenzhong1) [reply neural_speed/models/grok/grok.h:29]: The second scratch memory is less than the first and third temporary memories.

Just make sure the 2nd one uses the least and change the 1st and 3rd from 3072 to 4096, keeping the same as other models.

**Review by Zhenzhong1 — APPROVED** [review]

**Comment by intellinjun** [comment]
image
use mha fusion

**Comment by intellinjun** [comment]
image
q4g32cpint8 model  inference result

**Comment by intellinjun** [comment]
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/linjun1/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/linjun1/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">


model | precision | batch_size | input | output | cores | first_token | next_token | total_latency
-- | -- | -- | -- | -- | -- | -- | -- | --
grok314B | q4_j_b32_cint8 | 1 | 32 | 32 | 56 | 1976.39 | 381.64 | 13807.23
grok314B | q4_j_b32_cint8 | 1 | 1024 | 32 | 56 | 24366.39 | 384.4 | 36282.79
grok314B | q4_j_b128_cint8 | 1 | 32 | 32 | 56 | 728.79 | 283.63 | 9521.32
grok314B | q4_j_b128_cint8 | 1 | 1024 | 32 | 56 | 5942.93 | 289.87 | 14928.9
grok314B | q4_0 | 1 | 32 | 32 | 56 | 7432.41 | 407.29 | 20058.4
grok314B | q4_0 | 1 | 1024 | 32 | 56 | 110467.3 | 385.43 | 122415.7



</body>

</html>

**Pull request merged**